src/os/bluestore/KernelDevice.cc

```diff
@@ -64,23 +68,45 @@ int KernelDevice::_lock()
 int KernelDevice::open(const string& p)
 {
   path = p;
+  int i = 0;
   int r = 0;
   dout(1) << __func__ << " path " << path << dendl;
 
-  fd_direct = ::open(path.c_str(), O_RDWR | O_DIRECT | O_CLOEXEC);
-  BlkDev blkdev_direct(fd_direct);
-  if (fd_direct < 0) {
-    r = -errno;
-    derr << __func__ << " open got: " << cpp_strerror(r) << dendl;
-    return r;
+  for (i = 0; i < WRITE_LIFE_MAX; i++) {
+    int fd = ::open(path.c_str(), O_RDWR | O_DIRECT);
+    if (fd  < 0) {
+      r = -errno;
+      break;
+    }
+    fd_directs[i] = fd;
+
+    fd  = ::open(path.c_str(), O_RDWR | O_CLOEXEC);
+    if (fd  < 0) {
+      r = -errno;
+      break;
+    }
+    fd_buffereds[i] = fd;
   }
-  fd_buffered = ::open(path.c_str(), O_RDWR | O_CLOEXEC);
-  BlkDev blkdev_buffered(fd_buffered);
-  if (fd_buffered < 0) {
-    r = -errno;
-    derr << __func__ << " open got: " << cpp_strerror(r) << dendl;
-    goto out_direct;
+  if (i != WRITE_LIFE_MAX) {
+      derr << __func__ << " open got: " << cpp_strerror(r) << dendl;
+      goto out_fail;
   }
+
+  for (i = 0; i < WRITE_LIFE_MAX; i++) {
+    if (fcntl(fd_directs[i], F_SET_FILE_RW_HINT, &i) < 0) {
```

Comment:
Will this fail on old kernels that we care about?  Should we tolerate EOPNOTSUPP or similar?

Reply:
If failed on old kernel, we disable enable_wrt and later in choose_fd, we only use fd_directs/fd_buffered[WRITE_LIFE_NOT_SET].